### PR TITLE
Register `jakarta.json.spi.JsonProvider` service provider

### PR DESCRIPTION
**Pull request by zakkak**
`META-INF/services/jakarta.json.spi.JsonProvider` is accessed as a
resource by SmallRyeHealthReporter.

Relates to https://github.com/quarkusio/quarkus/issues/41995

### CODE REVIEW

**Review by gsmet — CHANGES_REQUESTED** [review]
Hmmm, shouldn't this be handled by the `quarkus-jsonp` extension?

**Comment by gsmet** [comment]
Yeah my guess is that it should be added here instead: https://github.com/quarkusio/quarkus/blob/main/extensions/jsonp/deployment/src/main/java/io/quarkus/jsonp/deployment/JsonpProcessor.java .

**Comment by zakkak** [comment]
> Are you sure? Because this output can be misleading: it will only display the dependency if not displayed earlier.

Hmm I didn't know about that. Thanks!

In that case you are right. Moving it.

**Comment by zakkak** [comment]
> Could you try with `./mvnw quarkus:dependency-tree` if things are better?

Nope, still doesn't show up.

```
[INFO] ├─ io.quarkus:quarkus-smallrye-health-deployment:999-SNAPSHOT (compile)
[INFO] │  ├─ io.quarkus:quarkus-smallrye-health-spi:999-SNAPSHOT (compile)
[INFO] │  ├─ io.quarkus:quarkus-smallrye-health:999-SNAPSHOT (compile)
[INFO] │  │  ├─ io.smallrye:smallrye-health:4.1.0 (compile)
[INFO] │  │  │  ├─ org.eclipse.microprofile.health:microprofile-health-api:4.0.1 (compile)
[INFO] │  │  │  ├─ io.smallrye:smallrye-health-api:4.1.0 (compile)
[INFO] │  │  │  └─ jakarta.json:jakarta.json-api:2.1.3 (compile)
[INFO] │  │  └─ io.smallrye:smallrye-health-provided-checks:4.1.0 (compile)
[INFO] │  └─ io.smallrye:smallrye-health-ui:4.1.0 (compile)
```

**Comment by gsmet** [comment]
OK, I'm pretty sure it's just an artifact of the tools so please add it to `quarkus-jsonp-deployment` and if it doesn't fix your issue we can discuss it further.

**Comment by zakkak** [comment]
Dine

**Review by xstefank — APPROVED** [review]

File: extensions/jsonp/deployment/src/main/java/io/quarkus/jsonp/deployment/JsonpProcessor.java [hunk 1]
```diff
@@ -5,16 +5,20 @@
...
     @BuildStep
     void build(BuildProducer<FeatureBuildItem> feature,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
-            BuildProducer<NativeImageResourceBundleBuildItem> resourceBundle) {
+            BuildProducer<NativeImageResourceBuildItem> serviceProviders) {
         reflectiveClass.produce(
                 ReflectiveClassBuildItem.builder(JsonProviderImpl.class.getName()).build());
+
+        // Register META-INF/services/jakarta.json.spi.JsonProvider as a resource, accessed by SmallRyeHealthReporter
+        serviceProviders.produce(new NativeImageResourceBuildItem("META-INF/services/jakarta.json.spi.JsonProvider"));
```
Comment (gsmet) [inline extensions/jsonp/deployment/src/main/java/io/quarkus/jsonp/deployment/JsonpProcessor.java:21]: Sorry I missed that earlier but is there a reason why you don't use `ServiceProviderBuildItem`? It seems more appropriate?

**Review by gsmet — CHANGES_REQUESTED** [review]
Let's clarify my question before merging (you might have a good reason to use this particular build item).

Also could you please update the title of the PR?

File: extensions/jsonp/deployment/src/main/java/io/quarkus/jsonp/deployment/JsonpProcessor.java (hunk 1 above)
Comment (zakkak) [reply extensions/jsonp/deployment/src/main/java/io/quarkus/jsonp/deployment/JsonpProcessor.java:21]: Good point, a `ServiceProviderBuildItem` without any providers just the job and seems more appropriate.

Thanks Guillaume!

**Review by gsmet — APPROVED** [review]

File: extensions/jsonp/deployment/src/main/java/io/quarkus/jsonp/deployment/JsonpProcessor.java [hunk 2]
```diff
@@ -5,16 +5,20 @@
...
     @BuildStep
     void build(BuildProducer<FeatureBuildItem> feature,
             BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
-            BuildProducer<NativeImageResourceBundleBuildItem> resourceBundle) {
+            BuildProducer<ServiceProviderBuildItem> serviceProviders) {
         reflectiveClass.produce(
                 ReflectiveClassBuildItem.builder(JsonProviderImpl.class.getName()).build());
+
+        // Register jakarta.json.spi.JsonProvider provider which is accessed by SmallRyeHealthReporter
+        serviceProviders.produce(new ServiceProviderBuildItem("jakarta.json.spi.JsonProvider"));
```
Comment (gastaldi) [inline extensions/jsonp/deployment/src/main/java/io/quarkus/jsonp/deployment/JsonpProcessor.java:21]:
```suggestion
        serviceProviders.produce(new ServiceProviderBuildItem("jakarta.json.spi.JsonProvider", "org.eclipse.parsson.JsonProviderImpl"));
```

**Comment by gsmet** [comment]
Actually, we did it entirely wrong from the start (and I mean when we created the extension). I fixed it by amending your commit. Things should be fine from all point of view now.

**Pull request merged**